src/compiler/checker.ts

```diff
@@ -12877,9 +12879,17 @@ namespace ts {
         }
 
         function getTupleElementType(type: Type, index: number) {
-            return isTupleType(type) ?
-                index < getLengthOfTupleType(type) ? type.typeArguments![index] : getRestTypeOfTupleType(type) :
-                getTypeOfPropertyOfType(type, "" + index as __String);
+            const propType = getTypeOfPropertyOfType(type, "" + index as __String);
+            if (propType) {
+                return propType;
+            }
+            if (everyType(type, isTupleType)) {
+                const restType = mapType(type, t => getRestTypeOfTupleType(<TupleTypeReference>t) || undefinedType);
+                if (restType !== undefinedType) {

```

Comment:
Yeah, same remark - this needs to use a unique marker and not the `undefinedType` to signal the `restType`'s return, as it prevents a legitimate `undefined` rest type from functioning.